In [1]:
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install(version = "3.10")
# BiocManager::install("randomForest")

In [2]:
setwd('/Users/alexis/University of North Carolina at Chapel Hill/Rager, Julia - CEMALB_DataAnalysisPM/Projects/P1002. Ecig Metabolomics NELF/P1002.3. Analyses/P1002.3.2. Smoking Status Prediction/Input')
Output = ('/Users/alexis/University of North Carolina at Chapel Hill/Rager, Julia - CEMALB_DataAnalysisPM/Projects/P1002. Ecig Metabolomics NELF/P1002.3. Analyses/P1002.3.1. Smoking Status Prediction/Output')
cur_date = "021922"

library(tidyverse)
library(caret)
library(e1071)
library(Hmisc)
library(glmnet)
library(pROC)
library(randomForest)

#reading in file
molecular_feature_df_w_Ecig_40 = read_csv("20220217_mf_qrilc_w40.csv") %>%
     rename(Smoking_Status = sample)
molecular_feature_df_wo_Ecig_40 = read_csv("20220217_mf_qrilc_wo40.csv") %>%
     rename(Smoking_Status = sample)
# ("021422_imputed_cytokine_data_w_Ecig40.csv") %>%
#     rename(Smoking_Status = X1)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.3     ✔ purrr   0.3.4
✔ tibble  3.0.6     ✔ dplyr   1.0.4
✔ tidyr   1.1.2     ✔ stringr 1.4.0
✔ readr   1.4.0     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()

Loading required package: lattice


Attaching package: ‘caret’


The following object is masked from ‘package:purrr’:

    lift


Loading required package: survival


Attaching package: ‘survival’


The following object is masked from ‘package:caret’:

    cluster


Loading required package: Formula


Attaching package: ‘Hmisc’


The following object is masked from ‘package:e1071’:

    impute


The following objects are masked from ‘package:dplyr’:

    src, summarize


The following objects are masked from ‘package:base’:

    format.pval, units


Loading required package: Matrix


Attaching package: ‘Mat

In [3]:
head(molecular_feature_df_w_Ecig_40)

Smoking_Status MF_2     MF_3     MF_4     MF_5     MF_6     MF_7     MF_11   
1 NS_070         19.86186 17.86943 19.21394 18.02809 17.46636 18.34174 18.41799
2 NS_110         21.08622 22.58769 21.05378 17.93584 17.33098 20.41479 20.37485
3 NS_143         13.12099 14.44659 13.46827 16.51009 14.93820 15.66299 15.72731
4 NS_152         21.60361 23.22761 21.99010 16.22547 17.59020 21.53908 21.14531
5 NS_187         19.43885 17.93876 19.68977 16.73360 17.88844 17.81758 17.64979
6 NS_196         22.53459 24.36662 22.77595 16.69956 16.38411 22.55912 17.30008
  MF_12    MF_13    ⋯ MF_62853 MF_63431 MF_63701 MF_64025 MF_64936 MF_65532
1 19.31361 19.46777 ⋯ 19.06974 19.10260 17.17255 16.92465 17.09794 19.06203
2 21.12876 21.01488 ⋯ 18.02560 17.82667 19.37426 16.23450 19.06421 18.06379
3 13.78698 15.18975 ⋯ 19.14095 19.02411 19.56175 17.17120 16.97302 19.49121
4 22.07857 21.97502 ⋯ 19.28751 19.54886 16.78848 16.46750 19.47635 19.44477
5 19.64269 19.32233 ⋯ 16.70042 17.85969 18.97406 18.18813 18.50336 19.19085
6 22.80146 22.09338 ⋯ 16.86999 15.82637 19.46467 16.97149 17.46699 16.32598
  MF_68286 MF_73340 MF_77165 MF_78976
1 19.13556 16.93182 17.55702 17.05593
2 19.39103 16.72112 19.27343 19.30881
3 17.09470 15.56214 19.55038 19.54912
4 16.99617 16.09531 19.28810 19.32788
5 19.41757 16.88993 19.35678 18.95550
6 15.62938 17.37207 15.84334 17.82000

In [4]:
#first making smoking status column binary denoted by 0 = non smoker and 1 = ecig user
molecular_feature_df_w_Ecig_40 = molecular_feature_df_w_Ecig_40 %>%
    #grepl is synonymous with "contains"
    #necessary to put it into factor and relevel, so algorithm understands which data is considered baseline
    mutate(Smoking_Status = relevel(factor(ifelse(grepl("Ecig", Smoking_Status), 1, 0)), ref = "0"))

molecular_feature_df_wo_Ecig_40 = molecular_feature_df_wo_Ecig_40 %>%
    #grepl is synonymous with "contains"
    #necessary to put it into factor and relevel, so algorithm understands which data is considered baseline
    mutate(Smoking_Status = relevel(factor(ifelse(grepl("Ecig", Smoking_Status), 1, 0)), ref = "0"))

head(molecular_feature_df_w_Ecig_40)

Smoking_Status MF_2     MF_3     MF_4     MF_5     MF_6     MF_7     MF_11   
1 0              19.86186 17.86943 19.21394 18.02809 17.46636 18.34174 18.41799
2 0              21.08622 22.58769 21.05378 17.93584 17.33098 20.41479 20.37485
3 0              13.12099 14.44659 13.46827 16.51009 14.93820 15.66299 15.72731
4 0              21.60361 23.22761 21.99010 16.22547 17.59020 21.53908 21.14531
5 0              19.43885 17.93876 19.68977 16.73360 17.88844 17.81758 17.64979
6 0              22.53459 24.36662 22.77595 16.69956 16.38411 22.55912 17.30008
  MF_12    MF_13    ⋯ MF_62853 MF_63431 MF_63701 MF_64025 MF_64936 MF_65532
1 19.31361 19.46777 ⋯ 19.06974 19.10260 17.17255 16.92465 17.09794 19.06203
2 21.12876 21.01488 ⋯ 18.02560 17.82667 19.37426 16.23450 19.06421 18.06379
3 13.78698 15.18975 ⋯ 19.14095 19.02411 19.56175 17.17120 16.97302 19.49121
4 22.07857 21.97502 ⋯ 19.28751 19.54886 16.78848 16.46750 19.47635 19.44477
5 19.64269 19.32233 ⋯ 16.70042 17.85969 18.97406 18.18813 18.50336 19.19085
6 22.80146 22.09338 ⋯ 16.86999 15.82637 19.46467 16.97149 17.46699 16.32598
  MF_68286 MF_73340 MF_77165 MF_78976
1 19.13556 16.93182 17.55702 17.05593
2 19.39103 16.72112 19.27343 19.30881
3 17.09470 15.56214 19.55038 19.54912
4 16.99617 16.09531 19.28810 19.32788
5 19.41757 16.88993 19.35678 18.95550
6 15.62938 17.37207 15.84334 17.82000

The goal of this analysis is to use molecular features [continuous] to predict smoking staus (dichotomous). This will be done using Support Vector Machine (SVM) based on a logistic regression model for classification and Random Forest.

# Random Forest

- an ensemble learning method operating by constructing a multitude of decision trees at training time, which uses multiple methods to obtain a better predictive performance and includes bagging and random forest
- algorithm uses a bootstrop aggregation, to reduce overfitting the training datset but only a subset of the features are used hence decorrelation of predictors

In [5]:
rf_classification = function(dataset, outcome, pred_outcome){
    #setting for reproducibility
    set.seed(12)
    #splitting data into training and testing sets
    dataset_index = createFolds(dataset[[outcome]], k = 5) #5 fold CV

    metrics = data.frame()
    variable_importance_df = data.frame()
    
    for (i in 1:length(dataset_index)){
        data_train = dataset[-dataset_index[[i]],]
        data_test = dataset[dataset_index[[i]],]

        ntree_values = c(50, 250, 500) #number of trees (not sure if we wanted something different)
        p = dim(dataset)[2] - 1 #number of variables in dataset
        mtry_values = c(p/2, sqrt(p), p)

        #will use ntree and mtry values to determine which combination yields the smallest MSE
        reg_rf_pred_tune = list()
        rf_OOB_errors = list()
        rf_error_df = data.frame()
        for (j in 1:length(ntree_values)){
            for (k in 1:length(mtry_values)){
                reg_rf_pred_tune[[k]] = randomForest(as.formula(paste0(outcome, "~.")), data = data_train, 
                                                     ntree = ntree_values[j], mtry = mtry_values[k])
                rf_OOB_errors[[k]] = data.frame("Tree Number" = ntree_values[j], "Variable Number" = mtry_values[k], 
                                       "OOB_errors" = reg_rf_pred_tune[[k]]$err.rate[ntree_values[j],1])
                rf_error_df = rbind(rf_error_df, rf_OOB_errors[[k]])
            }
        }

        #finding the lowest OOB error using best number of predictors at split and refitting OG tree
        best_oob_errors <- which(rf_error_df$OOB_errors == min(rf_error_df$OOB_errors))

        reg_rf <- randomForest(as.formula(paste0(outcome, "~.")), data = data_train,
        #need to go back and figure out how the number of trees/ variables make a difference even though
        #the errors are the same!!!!
                               ntree = rf_error_df$Tree.Number[min(best_oob_errors)], 
                               mtry = rf_error_df$Variable.Number[min(best_oob_errors)])

        #predicting on test set
        data_test[[pred_outcome]] = predict(reg_rf, newdata = data_test, type = "response")


        matrix = confusionMatrix(data = data_test[[pred_outcome]], reference = data_test[[outcome]], 
                                     positive = "1")

        #calculating AUC
        auc = auc(response = data_test[[outcome]], predictor = factor(data_test[[pred_outcome]], ordered = TRUE))
        #extracting accuracy, sens, spec, PPV to take mean later
        matrix_values = data.frame(t(c(matrix$byClass[11])), t(c(matrix$byClass[1:3])), auc)
        
        #extracting variable importance
        var_importance_values = data.frame(importance(reg_rf)) %>%
            rownames_to_column(var = "MF")
        variable_importance_df = rbind(variable_importance_df, var_importance_values)
   
        #adding values to df
        metrics = rbind(metrics, matrix_values)
    }
    
    #taking averages/sd 
    metrics = metrics %>%
        summarise(`Balanced Accuracy` = mean(Balanced.Accuracy), Sensitivity = mean(Sensitivity), 
              Specificity = mean(Specificity), PPV = mean(Pos.Pred.Value), AUC = mean(auc))
    
    variable_importance_df = variable_importance_df %>%
        group_by(MF) %>%
        summarise(MeanDecreaseGini = mean(MeanDecreaseGini)) %>%
        #sorting by most important variables
        arrange(-MeanDecreaseGini)
  
    #return training set, matrix and variable importance values
    return(list(data_train, metrics, variable_importance_df))

}

#calling fn
rf_values_w_Ecig_40 = rf_classification(molecular_feature_df_w_Ecig_40, "Smoking_Status", "pred_Smoking_Status")
rf_values_wo_Ecig_40 = rf_classification(molecular_feature_df_wo_Ecig_40, "Smoking_Status", "pred_Smoking_Status")

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases



In [7]:
#viewing results
rf_confusion_matrix = data.frame(Model = c(rep(c("Original RF"), times = 2)), Dataset = c("Ecig_40", "no_Ecig_40"),
                                 Kernel = c(rep(c(NA), times = 2)), rbind(rf_values_w_Ecig_40[[2]], 
                                                                          rf_values_wo_Ecig_40[[2]]))
rf_confusion_matrix

#viewing most significant features
head(rf_values_w_Ecig_40[[3]])
head(rf_values_wo_Ecig_40[[3]]) 

Model       Dataset    Balanced.Accuracy Sensitivity Specificity PPV      
1 Original RF Ecig_40    0.9047619         0.8428571   0.9666667   0.9600000
2 Original RF no_Ecig_40 0.9047619         0.8428571   0.9666667   0.9666667
  AUC      
1 0.9047619
2 0.9047619

MF       MeanDecreaseGini
1  MF_97    7.05424062      
2  MF_5     4.68378298      
3  MF_292   0.27705811      
4  MF_3345  0.24800949      
5  MF_1713  0.18760701      
6  MF_4803  0.16542196      
7  MF_7423  0.15562403      
8  MF_5782  0.15079241      
9  MF_7221  0.14144676      
10 MF_3061  0.14079651      
11 MF_5547  0.12181102      
12 MF_17380 0.12076106      
13 MF_4858  0.11412593      
14 MF_1216  0.10876958      
15 MF_1324  0.10385556      
16 MF_2428  0.10128069      
17 MF_4876  0.09343499      
18 MF_8500  0.08960516      
19 MF_6044  0.08937172      
20 MF_30084 0.08096736      
21 MF_6608  0.08048303      
22 MF_2213  0.07969302      
23 MF_2341  0.07918822      
24 MF_253   0.07836272      
25 MF_8580  0.07598179      
26 MF_1004  0.07224050      
27 MF_14169 0.07163861      
28 MF_10060 0.06986090      
29 MF_8524  0.06785672      
30 MF_610   0.06784015      
31 MF_9035  0.06662714      
32 MF_1161  0.06468369      
33 MF_1522  0.06100982      
34 MF_15476 0.06043130      
35 MF_17896 0.05959955      
36 MF_39377 0.05866950      
37 MF_579   0.05786044      
38 MF_692   0.05594819      
39 MF_621   0.05435496      
40 MF_7626  0.05259605      
41 MF_2218  0.05158218

MF       MeanDecreaseGini
1  MF_97    7.05972731      
2  MF_5     4.08153303      
3  MF_13899 0.29520675      
4  MF_4858  0.19183248      
5  MF_5547  0.19155348      
6  MF_292   0.18847797      
7  MF_17380 0.15940730      
8  MF_108   0.15200375      
9  MF_873   0.15093885      
10 MF_2213  0.14683360      
11 MF_8500  0.13820279      
12 MF_10060 0.13182703      
13 MF_8306  0.12860052      
14 MF_1324  0.12063404      
15 MF_4803  0.10907961      
16 MF_21211 0.08910105      
17 MF_7181  0.08716910      
18 MF_6044  0.08596256      
19 MF_5782  0.08577525      
20 MF_3345  0.08272908      
21 MF_8828  0.08088273      
22 MF_8524  0.08081984      
23 MF_19761 0.07670596      
24 MF_4418  0.07584827      
25 MF_658   0.07249396      
26 MF_22381 0.06926002      
27 MF_8523  0.06870212      
28 MF_2306  0.06853988      
29 MF_21694 0.06524823      
30 MF_2547  0.06471759      
31 MF_3118  0.06425386      
32 MF_6993  0.06406617      
33 MF_19175 0.06403132      
34 MF_1513  0.06299532      
35 MF_24263 0.06236058      
36 MF_10895 0.06034167      
37 MF_1017  0.06022756      
38 MF_11851 0.05899666      
39 MF_2852  0.05897188      
40 MF_1036  0.05761792      
41 MF_3061  0.05572643      
42 MF_27928 0.05420874      
43 MF_1713  0.05270328      
44 MF_6634  0.05142615      
45 MF_37437 0.05092769      
46 MF_11011 0.05087719

Rerunning RF model with noise variables to determine, which predictors are significant above the background noise. 

In [10]:
train_vars_noise_w_Ecig_40 = rf_values_w_Ecig_40[[1]]
train_vars_noise_wo_Ecig_40 = rf_values_wo_Ecig_40[[1]]

# Add random noise predictors as an additional method to evaluate model performance
# Adding a column that contains randomly shuffled values from one of the molecules; sampling with replacement
train_vars_noise_w_Ecig_40$noise1 = sample(train_vars_noise_w_Ecig_40[[colnames(train_vars_noise_w_Ecig_40[3])]], 
                                           replace = TRUE) 
train_vars_noise_w_Ecig_40$noise2 = sample(train_vars_noise_w_Ecig_40[[colnames(train_vars_noise_w_Ecig_40[4])]], 
                                           replace = TRUE)
train_vars_noise_w_Ecig_40$noise3 = sample(train_vars_noise_w_Ecig_40[[colnames(train_vars_noise_w_Ecig_40[5])]], 
                                           replace = TRUE)
train_vars_noise_w_Ecig_40$noise4 = sample(train_vars_noise_w_Ecig_40[[colnames(train_vars_noise_w_Ecig_40[6])]], 
                                           replace = TRUE)
train_vars_noise_w_Ecig_40$noise5 = sample(train_vars_noise_w_Ecig_40[[colnames(train_vars_noise_w_Ecig_40[7])]], 
                                           replace = TRUE)

train_vars_noise_wo_Ecig_40$noise1 = sample(train_vars_noise_wo_Ecig_40[[colnames(train_vars_noise_wo_Ecig_40[3])]], 
                                           replace = TRUE) 
train_vars_noise_wo_Ecig_40$noise2 = sample(train_vars_noise_wo_Ecig_40[[colnames(train_vars_noise_wo_Ecig_40[4])]], 
                                           replace = TRUE)
train_vars_noise_wo_Ecig_40$noise3 = sample(train_vars_noise_wo_Ecig_40[[colnames(train_vars_noise_wo_Ecig_40[5])]], 
                                           replace = TRUE)
train_vars_noise_wo_Ecig_40$noise4 = sample(train_vars_noise_wo_Ecig_40[[colnames(train_vars_noise_wo_Ecig_40[6])]], 
                                           replace = TRUE)
train_vars_noise_wo_Ecig_40$noise5 = sample(train_vars_noise_wo_Ecig_40[[colnames(train_vars_noise_wo_Ecig_40[7])]], 
                                           replace = TRUE)

head(train_vars_noise_w_Ecig_40)

Smoking_Status MF_2     MF_3     MF_4     MF_5     MF_6     MF_7     MF_11   
1 0              19.86186 17.86943 19.21394 18.02809 17.46636 18.34174 18.41799
2 0              21.08622 22.58769 21.05378 17.93584 17.33098 20.41479 20.37485
3 0              13.12099 14.44659 13.46827 16.51009 14.93820 15.66299 15.72731
4 0              19.43885 17.93876 19.68977 16.73360 17.88844 17.81758 17.64979
5 0              22.53459 24.36662 22.77595 16.69956 16.38411 22.55912 17.30008
6 0              21.50216 23.34553 21.53633 17.93372 24.26751 21.71412 21.32122
  MF_12    MF_13    ⋯ MF_65532 MF_68286 MF_73340 MF_77165 MF_78976 noise1  
1 19.31361 19.46777 ⋯ 19.06203 19.13556 16.93182 17.55702 17.05593 22.88241
2 21.12876 21.01488 ⋯ 18.06379 19.39103 16.72112 19.27343 19.30881 23.00581
3 13.78698 15.18975 ⋯ 19.49121 17.09470 15.56214 19.55038 19.54912 23.46020
4 19.64269 19.32233 ⋯ 19.19085 19.41757 16.88993 19.35678 18.95550 22.38033
5 22.80146 22.09338 ⋯ 16.32598 15.62938 17.37207 15.84334 17.82000 17.40963
6 21.66214 21.71152 ⋯ 16.71579 18.52501 17.61326 17.74147 16.76017 23.90655
  noise2   noise3   noise4   noise5  
1 20.68832 26.26236 17.50820 17.81758
2 22.14102 26.95046 17.88701 17.69319
3 22.77595 18.47581 21.08991 21.33890
4 19.94664 18.18218 20.96841 22.54351
5 13.46827 20.21918 26.11076 21.71412
6 21.58279 17.32305 16.42013 20.05850

In [13]:
#calling rf fn
noise_rf_values_w_Ecig_40 = rf_classification(train_vars_noise_w_Ecig_40, "Smoking_Status", "pred_Smoking_Status")
noise_rf_values_wo_Ecig_40 = rf_classification(train_vars_noise_wo_Ecig_40, "Smoking_Status", 
                                               "pred_Smoking_Status")

In [12]:
#viewing results
noise_rf_confusion_matrix = data.frame(Model = c(rep(c("RF w/ Noise"), times = 2)), 
                                       Dataset = c("Ecig_40", "no_Ecig_40"), Kernel = c(rep(c(NA), times = 2)),
                                       rbind(noise_rf_values_w_Ecig_40[[2]], noise_rf_values_wo_Ecig_40[[2]]))
noise_rf_confusion_matrix

#filtering for most significant features
head(noise_rf_values_w_Ecig_40[[3]])
head(noise_rf_values_wo_Ecig_40[[3]]) 

Model       Dataset    Kernel Balanced.Accuracy Sensitivity Specificity
1 RF w/ Noise Ecig_40    NA     0.8866667         0.8933333   0.88       
2 RF w/ Noise no_Ecig_40 NA     0.8750000         0.8800000   0.87       
  PPV       AUC      
1 0.9250000 0.8866667
2 0.9028571 0.8750000

MF       MeanDecreaseGini
1  MF_97    7.05424062      
2  MF_5     4.68378298      
3  MF_292   0.27705811      
4  MF_3345  0.24800949      
5  MF_1713  0.18760701      
6  MF_4803  0.16542196      
7  MF_7423  0.15562403      
8  MF_5782  0.15079241      
9  MF_7221  0.14144676      
10 MF_3061  0.14079651      
11 MF_5547  0.12181102      
12 MF_17380 0.12076106      
13 MF_4858  0.11412593      
14 MF_1216  0.10876958      
15 MF_1324  0.10385556      
16 MF_2428  0.10128069      
17 MF_4876  0.09343499      
18 MF_8500  0.08960516      
19 MF_6044  0.08937172      
20 MF_30084 0.08096736      
21 MF_6608  0.08048303      
22 MF_2213  0.07969302      
23 MF_2341  0.07918822      
24 MF_253   0.07836272      
25 MF_8580  0.07598179      
26 MF_1004  0.07224050      
27 MF_14169 0.07163861      
28 MF_10060 0.06986090      
29 MF_8524  0.06785672      
30 MF_610   0.06784015      
31 MF_9035  0.06662714      
32 MF_1161  0.06468369      
33 MF_1522  0.06100982      
34 MF_15476 0.06043130      
35 MF_17896 0.05959955      
36 MF_39377 0.05866950      
37 MF_579   0.05786044      
38 MF_692   0.05594819      
39 MF_621   0.05435496      
40 MF_7626  0.05259605      
41 MF_2218  0.05158218

MF       MeanDecreaseGini
1  MF_97    7.05972731      
2  MF_5     4.08153303      
3  MF_13899 0.29520675      
4  MF_4858  0.19183248      
5  MF_5547  0.19155348      
6  MF_292   0.18847797      
7  MF_17380 0.15940730      
8  MF_108   0.15200375      
9  MF_873   0.15093885      
10 MF_2213  0.14683360      
11 MF_8500  0.13820279      
12 MF_10060 0.13182703      
13 MF_8306  0.12860052      
14 MF_1324  0.12063404      
15 MF_4803  0.10907961      
16 MF_21211 0.08910105      
17 MF_7181  0.08716910      
18 MF_6044  0.08596256      
19 MF_5782  0.08577525      
20 MF_3345  0.08272908      
21 MF_8828  0.08088273      
22 MF_8524  0.08081984      
23 MF_19761 0.07670596      
24 MF_4418  0.07584827      
25 MF_658   0.07249396      
26 MF_22381 0.06926002      
27 MF_8523  0.06870212      
28 MF_2306  0.06853988      
29 MF_21694 0.06524823      
30 MF_2547  0.06471759      
31 MF_3118  0.06425386      
32 MF_6993  0.06406617      
33 MF_19175 0.06403132      
34 MF_1513  0.06299532      
35 MF_24263 0.06236058      
36 MF_10895 0.06034167      
37 MF_1017  0.06022756      
38 MF_11851 0.05899666      
39 MF_2852  0.05897188      
40 MF_1036  0.05761792      
41 MF_3061  0.05572643      
42 MF_27928 0.05420874      
43 MF_1713  0.05270328      
44 MF_6634  0.05142615      
45 MF_37437 0.05092769      
46 MF_11011 0.05087719

# Support Vector Machine
- supervised learning models that can predict continuous (regression) or grouped/dichotomous (classification) outcomes
- predictors by projecting them onto a high dimensional space and uses kernels to make the data more separable (unfortunately makes interpretability of model results more difficult)
- does a better job at handling a large number of predictors since p > n

In [40]:
#model errors are calculated to assess the predictive accuracy of the model
svm_classification = function(dataset, model, outcome, pred_outcome, gamma_values, cost_values, elsilon_values){
  
    #setting seed for reproducibility
    set.seed(12)
    
    #splitting data into training and testing sets
    dataset_index = createFolds(dataset[[outcome]], k = 5) #5 fold CV
    metrics = data.frame()
    for (i in 1:length(dataset_index)){
        data_train = dataset[-dataset_index[[i]],]
        data_test = dataset[dataset_index[[i]],]

        #now pruning parameters (based on the training dataset to prevent overfitting)
        svr_tune <- tune(svm, as.formula(paste0(outcome, "~.")), data = data_train, kernel = model, 
                       ranges = list(elsilon = elsilon_values, cost = cost_values, gamma = gamma_values)) 

        #choosing best model
        best_svm <- svr_tune$best.model

        #predicting with tuned parameters 
        data_test[[pred_outcome]] <- predict(best_svm, newdata = data_test, type = "response")

        matrix = confusionMatrix(data = data_test[[pred_outcome]], reference = data_test[[outcome]], 
                                 positive = "1")
 
        #calculating AUC
        auc = auc(response = data_test[[outcome]], predictor = factor(data_test[[pred_outcome]], ordered = TRUE))
        #extracting accuracy, sens, spec, PPV to take mean later
        matrix_values = data.frame(t(c(matrix$byClass[11])), t(c(matrix$byClass[1:3])), auc)
   
        #adding values to df
        metrics = rbind(metrics, matrix_values)
  }
  
  #taking averages/sd 
  metrics = metrics %>%
    summarise(`Balanced Accuracy` = mean(Balanced.Accuracy), Sensitivity = mean(Sensitivity), 
              Specificity = mean(Specificity), PPV = mean(Pos.Pred.Value), AUC = mean(auc))
  
  return(metrics)
}

In [ ]:
#calling fn
#linear kernel
svm_linear_w_Ecig_40 = svm_classification(molecular_feature_df_w_Ecig_40, "linear", "Smoking_Status", "pred_Smoking_Status",
                                            0.035, 1:5, seq(0,1,0.2))
svm_linear_wo_Ecig_40 = svm_classification(molecular_feature_df_wo_Ecig_40, "linear", "Smoking_Status", "pred_Smoking_Status",
                                            0.035, 1:5, seq(0,1,0.2))
#radial
svm_radial_w_Ecig_40 = svm_classification(molecular_feature_df_w_Ecig_40, "radial", "Smoking_Status", "pred_Smoking_Status",
                                             0.035, 1:5, seq(0,1,0.2))
svm_radial_wo_Ecig_40 = svm_classification(molecular_feature_df_wo_Ecig_40, "radial", "Smoking_Status", "pred_Smoking_Status",
                                             0.035, 1:5, seq(0,1,0.2))
#polynomial
#elison value is fixed at 0.1, but I left it so I didn't have to create a new function
svm_polynomial_w_Ecig_40 = svm_classification(molecular_feature_df_w_Ecig_40, "polynomial", "Smoking_Status", 
                                             "pred_Smoking_Status", 0.035, 1:5, 0.1)
svm_polynomial_wo_Ecig_40 = svm_classification(molecular_feature_df_wo_Ecig_40, "polynomial", "Smoking_Status", 
                                  "pred_Smoking_Status", 0.035, 1:5, 0.1)

In [13]:
#creating one combined df
svm_final_df = data.frame(rbind(svm_linear_w_Ecig_40, svm_radial_w_Ecig_40, svm_polynomial_w_Ecig_40,
                               svm_linear_wo_Ecig_40, svm_radial_wo_Ecig_40, svm_polynomial_wo_Ecig_40))

svm_final_df = data.frame(Model = c(rep(c("Original SVM"), times = 6)), Dataset = c(rep(c("Ecig_40"), times = 3), 
                            rep(c("no_Ecig_40"), times = 3)), Kernel = rep(c("linear", "radial", "polynomial"), 
                                                                           times = 2), svm_final_df)
svm_final_df

Dataset    Kernel     Balanced.Accuracy Sensitivity Specificity PPV      
1 Ecig_40    linear     0.7614286         0.8095238   0.7133333   0.8044444
2 Ecig_40    radial     0.5000000         1.0000000   0.0000000   0.5424242
3 Ecig_40    polynomial 0.6057143         0.9047619   0.3066667   0.6187013
4 no_Ecig_40 linear     0.8100000         0.8000000   0.8200000   0.8666667
5 no_Ecig_40 radial     0.5000000         1.0000000   0.0000000   0.5348485
6 no_Ecig_40 polynomial 0.6057143         0.9047619   0.3066667   0.6070996
  AUC      
1 0.7614286
2 0.5000000
3 0.6057143
4 0.8100000
5 0.5000000
6 0.6057143

> The linear kernel performed the best overall and the inclusion of the outlier didn't make much of a difference.

# Multicollinearity Variable Selection w/ SVM

- Determing if there are highly correlated variables that can be removed, because multicollinearity makes variables that should be significant appear insignificant

In [3]:
#creating correlation matrix
#the cut off can be arbitrary, but due to the high number of correlations, 0.85 became the cutoff
corr_matrix_w_Ecig_40 = rcorr(as.matrix(molecular_feature_df_w_Ecig_40[2:3873]), type = "spearman")
#creating df where I can easily view coefficents that are highly correlated (rho > 0.7 or rho < -0.7)
corr_matrix_df_w_Ecig_40 = data.frame(corr_matrix_w_Ecig_40$r) %>%
  rownames_to_column(var = "Variable1") %>%
  #needed to specificy package of melt for it to work
  reshape2::melt(variable = "Variable2", value.name = "Spearman Coefficient") %>%
  #also removed values = 1, because those were correlations between the same variable
  filter(abs(`Spearman Coefficient`) < 1 & abs(`Spearman Coefficient`) > 0.7) %>%
  arrange(desc(`Spearman Coefficient`)) %>%
  #removing duplicate spearman coefficient values and keeping all columns
  distinct(`Spearman Coefficient`, .keep_all = TRUE) 

corr_matrix_wo_Ecig_40 = rcorr(as.matrix(molecular_feature_df_wo_Ecig_40[2:3873]), type = "spearman")
#creating df where I can easily view coefficents that are highly correlated (rho > 0.7 or rho < -0.7)
corr_matrix_df_wo_Ecig_40 = data.frame(corr_matrix_wo_Ecig_40$r) %>%
  rownames_to_column(var = "Variable1") %>%
  #needed to specificy package of melt for it to work
  reshape2::melt(variable = "Variable2", value.name = "Spearman Coefficient") %>%
  #also removed values = 1, because those were correlations between the same variable
  filter(abs(`Spearman Coefficient`) < 1 & abs(`Spearman Coefficient`) > 0.7) %>%
  arrange(desc(`Spearman Coefficient`)) %>%
  #removing duplicate spearman coefficient values and keeping all columns
  distinct(`Spearman Coefficient`, .keep_all = TRUE) 

head(corr_matrix_df_w_Ecig_40)

Using Variable1 as id variables

Using Variable1 as id variables



Variable1 Variable2 Spearman Coefficient
1 MF_586    MF_7      0.9987726           
2 MF_124    MF_34     0.9984220           
3 MF_516    MF_7      0.9974284           
4 MF_301    MF_133    0.9973115           
5 MF_54     MF_4      0.9969024           
6 MF_2762   MF_974    0.9963764

In [4]:
corr_matrix_df_w_Ecig_40
corr_matrix_df_wo_Ecig_40

Variable1 Variable2 Spearman Coefficient
1    MF_586    MF_7      0.9987726           
2    MF_124    MF_34     0.9984220           
3    MF_516    MF_7      0.9974284           
4    MF_301    MF_133    0.9973115           
5    MF_54     MF_4      0.9969024           
6    MF_2762   MF_974    0.9963764           
7    MF_586    MF_516    0.9960257           
8    MF_3375   MF_301    0.9945061           
9    MF_2827   MF_277    0.9942139           
10   MF_1409   MF_1136   0.9940970           
11   MF_1655   MF_1290   0.9935126           
12   MF_125    MF_13     0.9932203           
13   MF_3566   MF_516    0.9925774           
14   MF_3566   MF_7      0.9923437           
15   MF_3375   MF_133    0.9921099           
16   MF_1987   MF_289    0.9920514           
17   MF_14671  MF_7303   0.9915254           
18   MF_5068   MF_1409   0.9902981           
19   MF_4532   MF_1136   0.9901812           
20   MF_2005   MF_139    0.9898890           
21   MF_3566   MF_586    0.9897136           
22   MF_1655   MF_1409   0.9895383           
23   MF_4532   MF_1655   0.9894798           
24   MF_1430   MF_505    0.9887785           
25   MF_3375   MF_936    0.9881940           
26   MF_93     MF_4      0.9880771           
27   MF_1537   MF_516    0.9877265           
28   MF_4532   MF_1290   0.9876096           
29   MF_93     MF_54     0.9874927           
30   MF_1655   MF_1136   0.9873758           
⋮    ⋮         ⋮         ⋮                   
4300 MF_3039   MF_2028   -0.7816482          
4301 MF_23978  MF_18089  -0.7819404          
4302 MF_34980  MF_18089  -0.7825833          
4303 MF_12378  MF_5163   -0.7859731          
4304 MF_3039   MF_133    -0.7894798          
4305 MF_18411  MF_18089  -0.7900058          
4306 MF_3375   MF_3039   -0.7907072          
4307 MF_18089  MF_10520  -0.7926943          
4308 MF_18089  MF_73     -0.7943308          
4309 MF_20283  MF_17636  -0.7947399          
4310 MF_3039   MF_301    -0.7965517          
4311 MF_45154  MF_18089  -0.7981882          
4312 MF_17636  MF_15541  -0.7983635          
4313 MF_18089  MF_4952   -0.8018118          
4314 MF_17636  MF_15381  -0.8062537          
4315 MF_17636  MF_73     -0.8084746          
4316 MF_17636  MF_15126  -0.8171829          
4317 MF_17636  MF_10520  -0.8191701          
4318 MF_18089  MF_152    -0.8196961          
4319 MF_18089  MF_545    -0.8283460          
4320 MF_45154  MF_17636  -0.8296902          
4321 MF_8895   MF_5167   -0.8330216          
4322 MF_3178   MF_3039   -0.8351257          
4323 MF_18411  MF_17636  -0.8366452          
4324 MF_18089  MF_15     -0.8437171          
4325 MF_34980  MF_17636  -0.8476914          
4326 MF_17636  MF_15     -0.8661601          
4327 MF_17636  MF_4952   -0.8905319          
4328 MF_17636  MF_152    -0.9005845          
4329 MF_17636  MF_545    -0.9011105

Variable1 Variable2 Spearman Coefficient
1    MF_586    MF_7      0.9987696           
2    MF_124    MF_34     0.9986465           
3    MF_516    MF_7      0.9973546           
4    MF_301    MF_133    0.9971700           
5    MF_54     MF_4      0.9969239           
6    MF_2762   MF_974    0.9963702           
7    MF_586    MF_516    0.9958165           
8    MF_1409   MF_1136   0.9945246           
9    MF_3375   MF_301    0.9941555           
10   MF_2827   MF_277    0.9939709           
11   MF_1655   MF_1290   0.9936633           
12   MF_14671  MF_7303   0.9936018           
13   MF_4532   MF_1409   0.9929866           
14   MF_3566   MF_516    0.9921868           
15   MF_3566   MF_7      0.9920637           
16   MF_3375   MF_133    0.9915716           
17   MF_1987   MF_289    0.9911409           
18   MF_125    MF_13     0.9909564           
19   MF_4532   MF_1136   0.9907103           
20   MF_5068   MF_1409   0.9905872           
21   MF_1655   MF_1409   0.9896029           
22   MF_4532   MF_1655   0.9893568           
23   MF_1430   MF_505    0.9892338           
24   MF_3566   MF_586    0.9891722           
25   MF_93     MF_4      0.9889877           
26   MF_2005   MF_139    0.9889261           
27   MF_93     MF_54     0.9880648           
28   MF_1003   MF_856    0.9876342           
29   MF_5068   MF_1136   0.9875112           
30   MF_2346   MF_2139   0.9873881           
⋮    ⋮         ⋮         ⋮                   
4087 MF_20283  MF_18089  -0.7789535          
4088 MF_3039   MF_936    -0.7790766          
4089 MF_3039   MF_2028   -0.7804916          
4090 MF_3039   MF_133    -0.7805531          
4091 MF_34980  MF_18089  -0.7810452          
4092 MF_3375   MF_3039   -0.7842444          
4093 MF_13965  MF_1001   -0.7851057          
4094 MF_18089  MF_15126  -0.7862746          
4095 MF_18089  MF_73     -0.7884893          
4096 MF_3039   MF_301    -0.7915654          
4097 MF_18411  MF_18089  -0.7929804          
4098 MF_45154  MF_18089  -0.7942108          
4099 MF_20283  MF_17636  -0.8019010          
4100 MF_17636  MF_73     -0.8022701          
4101 MF_18089  MF_10520  -0.8038082          
4102 MF_17636  MF_15381  -0.8040543          
4103 MF_8895   MF_5167   -0.8063306          
4104 MF_18089  MF_4952   -0.8076225          
4105 MF_18089  MF_152    -0.8126673          
4106 MF_17636  MF_10520  -0.8220185          
4107 MF_17636  MF_15126  -0.8240487          
4108 MF_18411  MF_17636  -0.8265711          
4109 MF_18089  MF_545    -0.8302009          
4110 MF_3178   MF_3039   -0.8357378          
4111 MF_18089  MF_15     -0.8417669          
4112 MF_17636  MF_15     -0.8525947          
4113 MF_34980  MF_17636  -0.8648374          
4114 MF_17636  MF_152    -0.8721585          
4115 MF_17636  MF_4952   -0.8892614          
4116 MF_17636  MF_545    -0.9034114

In [8]:
#determining how many times each feature occurs
correlation_occurences1_w_Ecig_40 = corr_matrix_df_w_Ecig_40 %>%
    group_by(Variable1) %>%
    count() %>%
    arrange(-n) %>%
    filter(n > 20)

#correlation_occurences1

correlation_occurences2_w_Ecig_40 = corr_matrix_df_w_Ecig_40 %>%
    group_by(Variable2) %>%
    count() %>%
    arrange(-n) %>%
    filter(n > 20)

#correlation_occurences2

all_corr_occurences = unique(c(correlation_occurences1$Variable1, as.character(correlation_occurences2$Variable2)))

In [9]:
all_corr_occurences

character(0)

In [10]:
#removing molecular features (MF) that correlate most often
`%notin%` <- Negate(`%in%`)
penalized_dataset = molecular_feature_df[,colnames(molecular_feature_df) %notin% all_corr_occurences]

In [13]:
#calling fn
#linear kernel
svm_linear = svm_classification(penalized_dataset, "linear", "Smoking_Status", "pred_Smoking_Status",
                                             0.035, 1:5, seq(0,1,0.2))
#radial
svm_radial = svm_classification(penalized_dataset, "radial", "Smoking_Status", "pred_Smoking_Status",
                                             0.035, 1:5, seq(0,1,0.2))
#polynomial
#elison value is fixed at 0.1, but I left it so I didn't have to create a new function
svm_polynomial = svm_classification(penalized_dataset, "polynomial", "Smoking_Status", 
                                             "pred_Smoking_Status", 0.035, 1:5, 0.1)


#creating one combined df
svm_penalized_final_df = data.frame(rbind(svm_linear, svm_radial, svm_polynomial))
svm_penalized_final_df = data.frame(Kernel = c("linear", "radial", "polynomial"), svm_penalized_final_df)
svm_penalized_final_df

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls > cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls > cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls > cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting leve

Kernel     Balanced.Accuracy Sensitivity Specificity PPV       AUC      
1 linear     0.4804762         0.5476190   0.4133333   0.5018254 0.5766667
2 radial     0.5252381         0.4904762   0.5600000   0.6000000 0.5919048
3 polynomial 0.4528571         0.4523810   0.4533333         NaN 0.4528571

# RF Variable Selection w/ SVM

Calculating the average noise rank to determine, which predictors ranked higher than the average noise rank. Then filtering the original datasets for those predictors and rerunning SVM.

In [14]:
#calculating avg noise rank out of all variables
noise_importance_values_w_Ecig_40 = noise_rf_values_w_Ecig_40[[3]] %>%
    column_to_rownames(var = "MF")
noise_importance_values_wo_Ecig_40 = noise_rf_values_wo_Ecig_40[[3]] %>%
    column_to_rownames(var = "MF")

noise1rank = which(rownames(noise_importance_values_w_Ecig_40) == "noise1")/nrow(noise_importance_values_w_Ecig_40)
noise2rank = which(rownames(noise_importance_values_w_Ecig_40) == "noise2")/nrow(noise_importance_values_w_Ecig_40)
noise3rank = which(rownames(noise_importance_values_w_Ecig_40) == "noise3")/nrow(noise_importance_values_w_Ecig_40)
noise4rank = which(rownames(noise_importance_values_w_Ecig_40) == "noise4")/nrow(noise_importance_values_w_Ecig_40)
noise5rank = which(rownames(noise_importance_values_w_Ecig_40) == "noise5")/nrow(noise_importance_values_w_Ecig_40)

#calculating avg mean decrease gini or background noise importance value
avg_noise_rank_w_Ecig_40 = (noise1rank + noise2rank + noise3rank + noise4rank + noise5rank)/5
avg_noise_rank_w_Ecig_40

noise1rank = which(rownames(noise_importance_values_wo_Ecig_40) == "noise1")/nrow(noise_importance_values_wo_Ecig_40)
noise2rank = which(rownames(noise_importance_values_wo_Ecig_40) == "noise2")/nrow(noise_importance_values_wo_Ecig_40)
noise3rank = which(rownames(noise_importance_values_wo_Ecig_40) == "noise3")/nrow(noise_importance_values_wo_Ecig_40)
noise4rank = which(rownames(noise_importance_values_wo_Ecig_40) == "noise4")/nrow(noise_importance_values_wo_Ecig_40)
noise5rank = which(rownames(noise_importance_values_wo_Ecig_40) == "noise5")/nrow(noise_importance_values_wo_Ecig_40)

avg_noise_rank_wo_Ecig_40 = (noise1rank + noise2rank + noise3rank + noise4rank + noise5rank)/5
avg_noise_rank_wo_Ecig_40

[1] 0.5411916

[1] 0.8350271

In [35]:
for (i in 1:length(noise_importance_values_w_Ecig_40$MeanDecreaseGini)){
    #if the row number (index)/# of variables is less than the avg noise rank the loop will continue
    if (i/length(noise_importance_values_w_Ecig_40$MeanDecreaseGini) > avg_noise_rank_w_Ecig_40){
        last_significant_rank_w_Ecig_40 = i - 1
        break
    }
}
last_significant_rank_w_Ecig_40

for (i in 1:length(noise_importance_values_wo_Ecig_40$MeanDecreaseGini)){
    #if the row number (index)/# of variables is less than the avg noise rank the loop will continue
    if (i/length(noise_importance_values_wo_Ecig_40$MeanDecreaseGini) > avg_noise_rank_wo_Ecig_40){
        last_significant_rank_wo_Ecig_40 = i - 1
        break
    }
}
last_significant_rank_wo_Ecig_40

[1] 2098

[1] 3237

Therefore, the dataset containing Ecig_40 contains 2,098 MFs above background, while the dataset without the outlier contains 3,237. This means that the outlier dampened the significance of about 1,000 MF just by being included in the dataset. 

In [50]:
#now getting those significant MFs
significant_MFs_w_Ecig_40 = rownames(noise_importance_values_w_Ecig_40)[1:last_significant_rank_w_Ecig_40]
significant_MFs_wo_Ecig_40 = rownames(noise_importance_values_wo_Ecig_40)[1:last_significant_rank_wo_Ecig_40]

#now subsetting the original dfs
rf_subset_MF_df_w_Ecig_40 = cbind(Smoking_Status = molecular_feature_df_w_Ecig_40$Smoking_Status, molecular_feature_df_w_Ecig_40[,colnames(molecular_feature_df_w_Ecig_40) %in% significant_MFs_w_Ecig_40])
rf_subset_MF_df_wo_Ecig_40 = cbind(Smoking_Status = molecular_feature_df_wo_Ecig_40$Smoking_Status, molecular_feature_df_wo_Ecig_40[,colnames(molecular_feature_df_wo_Ecig_40) %in% significant_MFs_wo_Ecig_40])
head(rf_subset_MF_df_w_Ecig_40)

Smoking_Status MF_5     MF_11    MF_12    MF_13    MF_14    MF_15    MF_17   
1 0              18.02809 18.41799 19.31361 19.46777 20.80616 18.72405 20.11458
2 0              17.93584 20.37485 21.12876 21.01488 19.71124 19.17696 19.82979
3 0              16.51009 15.72731 13.78698 15.18975 18.19551 18.84322 18.65454
4 0              16.22547 21.14531 22.07857 21.97502 17.30567 19.61857 18.35221
5 0              16.73360 17.64979 19.64269 19.32233 21.33752 19.56012 20.23552
6 0              16.69956 17.30008 22.80146 22.09338 19.51343 19.39161 20.72501
  MF_18    MF_21    ⋯ MF_53619 MF_54327 MF_54613 MF_56870 MF_57989 MF_61492
1 18.49551 18.42766 ⋯ 18.26770 16.83157 17.55214 19.09192 19.07769 18.88772
2 19.80685 22.61087 ⋯ 19.82107 16.89255 19.40468 19.14730 17.29091 15.81175
3 17.52149 18.26920 ⋯ 18.97545 19.18696 19.22546 18.95185 19.30798 18.98001
4 17.14531 22.47085 ⋯ 19.31315 18.69526 18.74873 19.40897 19.53008 17.50812
5 18.70226 21.53982 ⋯ 19.69692 19.06106 19.32826 19.12974 19.11060 17.61151
6 21.45596 21.96066 ⋯ 18.53569 17.74897 18.41935 19.60590 13.81876 19.24975
  MF_62137 MF_62282 MF_64936 MF_78976
1 17.47760 18.83525 17.09794 17.05593
2 16.35027 19.31905 19.06421 19.30881
3 15.64893 16.80361 16.97302 19.54912
4 17.78108 19.06662 19.47635 19.32788
5 17.97428 19.13998 18.50336 18.95550
6 16.06782 19.19848 17.46699 17.82000

In [52]:
#calling fn
#linear kernel
subset_svm_linear_w_Ecig_40 = svm_classification(rf_subset_MF_df_w_Ecig_40, "linear", "Smoking_Status", "pred_Smoking_Status",
                                            0.035, 1:5, seq(0,1,0.2))
subset_svm_linear_wo_Ecig_40 = svm_classification(rf_subset_MF_df_wo_Ecig_40, "linear", "Smoking_Status", "pred_Smoking_Status",
                                            0.035, 1:5, seq(0,1,0.2))
#radial
subset_svm_radial_w_Ecig_40 = svm_classification(rf_subset_MF_df_w_Ecig_40, "radial", "Smoking_Status", "pred_Smoking_Status",
                                             0.035, 1:5, seq(0,1,0.2))
subset_svm_radial_wo_Ecig_40 = svm_classification(rf_subset_MF_df_wo_Ecig_40, "radial", "Smoking_Status", "pred_Smoking_Status",
                                             0.035, 1:5, seq(0,1,0.2))
#polynomial
subset_svm_polynomial_w_Ecig_40 = svm_classification(rf_subset_MF_df_w_Ecig_40, "polynomial", "Smoking_Status", 
                                             "pred_Smoking_Status", 0.035, 1:5, 0.1)
subset_svm_polynomial_wo_Ecig_40 = svm_classification(rf_subset_MF_df_wo_Ecig_40, "polynomial", "Smoking_Status", 
                                  "pred_Smoking_Status", 0.035, 1:5, 0.1)

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting levels: control = 0, case = 1

Setting direction: controls < cases

Setting leve

In [53]:
#creating one combined df
subset_svm_final_df = data.frame(rbind(subset_svm_linear_w_Ecig_40, subset_svm_radial_w_Ecig_40, 
                                       subset_svm_polynomial_w_Ecig_40,subset_svm_linear_wo_Ecig_40, 
                                       subset_svm_radial_wo_Ecig_40, subset_svm_polynomial_wo_Ecig_40))

subset_svm_final_df = data.frame(Model = c(rep(c("SVM w/ RF Var Selection"), times = 6)), 
                                 Dataset = c(rep(c("Ecig_40"), times = 3), rep(c("no_Ecig_40"), times = 3)), 
                                 Kernel = rep(c("linear", "radial", "polynomial"), times = 2), 
                                 subset_svm_final_df)
subset_svm_final_df

Model                   Dataset    Kernel     Balanced.Accuracy Sensitivity
1 SVM w/ RF Var Selection Ecig_40    linear     0.7614286         0.8095238  
2 SVM w/ RF Var Selection Ecig_40    radial     0.5000000         1.0000000  
3 SVM w/ RF Var Selection Ecig_40    polynomial 0.6304762         0.8476190  
4 SVM w/ RF Var Selection no_Ecig_40 linear     0.7957143         0.8047619  
5 SVM w/ RF Var Selection no_Ecig_40 radial     0.5000000         1.0000000  
6 SVM w/ RF Var Selection no_Ecig_40 polynomial 0.6557143         0.9047619  
  Specificity PPV       AUC      
1 0.7133333   0.8044444 0.7614286
2 0.0000000   0.5424242 0.5000000
3 0.4133333   0.6806061 0.6304762
4 0.7866667   0.8230952 0.7957143
5 0.0000000   0.5348485 0.5000000
6 0.4066667   0.6461905 0.6557143

In [ ]:
#all results
final_df = rbind(rf_confusion_matrix, svm_final_df, svm_penalized_final_df, subset_svm_final_df)
final_df